src/GafferScene/Shader.cpp

```diff
@@ -455,11 +455,20 @@ class Shader::NetworkBuilder
 			}
 			else
 			{
+				// Even though we have an input connection, still store the value of the parameter on
+				// this node.  This allows us to keep track of the type of the connection, and is more
+				// aligned with how other shader network storage formats, like USD, work.
+				if( IECore::DataPtr value = static_cast<const Shader *>( parameter->node() )->parameterValue( parameter ) )
+				{
+					shader->parameters()[parameterName] = value;
+				}
```

Comment:
This change needs to be reflected in `hashParameter()` too. It feels like maybe this could also be simplified a fair bit now that both sides of the conditional are generating a parameter value _and_ processing connections in some way. Maybe we can refactor so that we don't need the conditional at all?

Reply:
Cleaned up a bit and fixed hashParameter in 5728fe7

I've also done a bit more of a refactor in 3cbf483, that feels a little simpler to me in general?

Reply:
The test failure unfortunately suggests an actual problem.  In our conversation this morning, if I recall it correctly, you said that when writing out a value on a parameter, we should always write out the local plug value.  If it's driven by an output, we make a real conneciton.  If that plug value is driven by something that isn't an output plug, then Gaffer's evaluation should provide an appropriate value for it, and this makes sure the value is actually of the right type ( as opposed to the value of the source plug, which could be a different type ).

Based on this reasoning, I did my refactor, and was feeling pretty happy with it in terms of cleanliness - I was able to have the parameter value always driven by the local plug value instead of being in the conditional, like you wanted.  But the test cases point out that I missed an aspect here:  disabled shaders are supposed to pass through a value based on correspondingInput, and the compute on a shader does not do this passthrough.

My first thought is that maybe the compute on a shader should return a passthrough value when its disabled, but metadata isn't supposed to affect computes, and that could have other weird effects.

Maybe I should just go back to the conditional that sometimes uses effectiveParameter that I started with ... though your point about it being wrong that this could pick up the wrong plug type stands.